packages/protocol/contracts/signal/SignalService.sol

```diff
@@ -35,18 +35,28 @@ import "./ISignalService.sol";
 contract SignalService is AuthorizableContract, ISignalService {
     using SafeCast for uint256;
 
+    /// @dev Meaning, for this chainId A, signal service contract B, in given snapshot world state
+    /// root 'C', then here is a storage proof for that signal service.
+    mapping(
+        uint256 signalServiceChainId
+            => mapping(
+                address signalServiceAddress
+                    => mapping(bytes32 worldStateRoot => bytes signalStorageRoot)
```

Comment:
Do store the `signalStorageRoot` directly as a `bytes32` because that only takes up a single slot. Storing it as `bytes` requires two storage slots (the length of the array (32 bytes) + the actual storage root (32 bytes)). You can just compare against `bytes32(0)` for the non-existing case.